aws/middleware/user_agent_test.go

```diff
@@ -231,6 +231,71 @@ func TestAddUserAgentKeyValue(t *testing.T) {
 	}
 }
 
+func TestAddUserAgentFeature(t *testing.T) {
+	restoreEnv := clearEnv()
+	defer restoreEnv()
+
+	cases := map[string]struct {
+		Features []UserAgentFeature
+		Expect   string
+	}{
+		"none": {
+			Features: []UserAgentFeature{},
+			Expect:   expectedAgent,
+		},
+		"one": {
+			Features: []UserAgentFeature{
+				UserAgentFeatureWaiter,
+			},
+			Expect: "m/B " + expectedAgent,
+		},
+		"two": {
+			Features: []UserAgentFeature{
+				UserAgentFeatureRetryModeAdaptive, // ensure stable order, and idempotent
```

Comment:
[nit] not sure if I'd call it stable order since this maps to a set at the end, and we end up sorting the feature set